Add new CLI '--check' option

roman-petrov opened the pull request:
Closes #41

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/format?pullRequest=48) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/format?pullRequest=48) before we can accept your contribution.<br/><br/>:x: roman-petrov [sign now](https://cla.dotnetfoundation.org/dotnet/format?pullRequest=48)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/format?pullRequest=48) it.</sub>

roman-petrov commented:
I just completed this PR and the implementation seems to work fine. I just added new `--check` option / flag. When this flag is specified `dotnet-format` CLI terminates with `1` exit code if any files in workspace formatted or should be formatted.

I updated existing unit tests according to my implementation. I used to following batch script to check if CLI application works as expected:

```bash
dotnet <dotnet-format.dll path> -v diag --check
if %ERRORLEVEL% GEQ 1 echo ERROR
if %ERRORLEVEL% EQU 0 echo OK
```

Now I am just removing `[WIP]` label and will wait for reviewer.

sharwell commented on src/Program.cs:
```diff
@@ -32,13 +32,14 @@ private static async Task<int> Main(string[] args)
         }
 
-        public static async Task<int> Run(string workspace, string verbosity, bool dryRun, IConsole console = null)
+        public static async Task<int> Run(string workspace, string verbosity, bool dryRun, bool check, IConsole console = null)
```
💡 Rather than pass this argument, we could _always_ simply return the number of files formatted from this method. The caller could then determine what to do with that number.

sharwell replied:
💭 Might need to use negative codes for errors, and positive codes for successful finding of number of files that need to be formatted. Otherwise the caller would not know whether 1 means error or no-error when `--check` is not given by the user.

roman-petrov replied:
> 💡 Rather than pass this argument, we could _always_ simply return the number of files formatted from this method. The caller could then determine what to do with that number.

Currently as I understand this method is being called by `CommandLineBuilder` command in code above:

`new Command("dotnet-format", 
handler: CommandHandler.Create(typeof(Program).GetMethod(nameof(Run))))`

New '--check' option is being passed in the same manner as all other options. So currently its not clear for me how to make the change you described, could you please advice?...

roman-petrov replied:
> 💭 Might need to use negative codes for errors, and positive codes for successful finding of number of files that need to be formatted. Otherwise the caller would not know whether 1 means error or no-error when `--check` is not given by the user.

Should we make some changes in this PR?..

JoeRobich replied:
@roman-petrov I think the current implementation is probably what is expected. If we need to make a change, we can do it later.

sharwell commented on README.md:
```diff
@@ -44,6 +44,7 @@ Options:
                      the current directory for one.
   -v, --verbosity    Set the verbosity level. Allowed values are q[uiet], m[inimal], n[ormal], d[etailed], and
                      diag[nostic]
+  --check            Terminate with non-zero exit code if any files formatted.
```
❔ Why not make this "the number of files formatted"?

roman-petrov replied:
In current implementation if `--check` option specified the application exits with `1` exit code, I did this like specified in `Prettier` documentation <https://prettier.io/docs/en/cli.html#check>. 

Do like to change the implementation to make `exit code=<number of files formetted>` when `--check` option is specified?

sharwell replied:
> Do like to change the implementation to make exit code=number of files formatted...?

I would not be a primary consumer of this feature, so other people with more experience using a `--check` option would need to answer your question. Your response will point those users to related documentation though which is a nice help. 😄

roman-petrov replied:
So lets leave it as it is or change is still needed?:)

jmarolf replied:
@roman-petrov 	I think the design that we doscussed for this is fine.

jmarolf commented:
@roman-petrov looks good to me.  I'll wait for @JoeRobich 	's signoff but could you add a test for this please?

JoeRobich commented on src/Program.cs:
```diff
@@ -81,13 +82,14 @@ public static async Task<int> Run(string workspace, string verbosity, bool dryRu
                     workspacePath,
                     isSolution,
                     logAllWorkspaceWarnings: logLevel == LogLevel.Trace,
                     saveFormattedFiles: !dryRun,
```
Does running a `--check` imply that `dryRun` should also be true or would you expect `--check` to update files on disk?

roman-petrov replied:
The `--check` option will affect only program **exit code**. All other options, even if combined with `--check` options, will continue to work in the same manner as worked without my PR. I just tried to make minimal working implementation in my PR which will solve continuous integration scenarios.

roman-petrov commented:
> @roman-petrov looks good to me. I'll wait for @JoeRobich 's signoff but could you add a test for this please?
In my PR I modified existing unit tests by adding cased that check `formatResult.FileCount`, `formatResult.FilesFormatted`. 

To test new logic that program exit code equal to `1` when `--check` option is used I will have to write a unit test for `Program::Run` method, which takes 'IConsole' as last parameter. C# currently is not a primary language I work with, so can you please give an advice how to add good test for this?..

I can do some refactoring to test new logic but I wonder it might be too many changes.

JoeRobich commented:
@roman-petrov I think the easiest way to make that code testable would be to pull the logic that determines the final exit code into a static function. You could then test that logic in isolation by passing in different combinations of `check` and `formatResult`.

roman-petrov commented:
> @roman-petrov I think the easiest way to make that code testable would be to pull the logic that determines the final exit code into a static function. You could then test that logic in isolation by passing in different combinations of `check` and `formatResult`.

Thank you, I was thinking about the same :) I added some tests, please take a look.

JoeRobich commented on tests/ProgramTests.cs:
```diff
@@ -0,0 +1,44 @@
+                ExitCode = 42,
+                FilesFormatted = 0
+            };
+            var exitCode = Program.GetExitCode(formatResult, true);
```
Use named arguments when passing boolean literals.
```suggestion
            var exitCode = Program.GetExitCode(formatResult, check: true);
```

JoeRobich commented on tests/ProgramTests.cs:
```diff
@@ -0,0 +1,44 @@
+            {
+                ExitCode = 42
+            };
+            var exitCode = Program.GetExitCode(formatResult, false);
```
Use named arguments when passing boolean literals.
```suggestion
            var exitCode = Program.GetExitCode(formatResult, check: false);
```

roman-petrov commented:
Is there anything needed to change or this PR can be merged?

JoeRobich commented:
@roman-petrov No changes needed. Thanks for contributing!

roman-petrov commented:
@JoeRobich thanks all your help!

robinmanuelthiel commented:
Thanks @roman-petrov! Needed exactly this! <3

roman-petrov commented:
@robinmanuelthiel Welcome! 😃

JoeRobich commented:
@boukeversteegh Starting with net6.0 SDK dotnet-format ships with the CLI and there is no need to install it as a tool. Along with this change the `--check` option became `--verify-no-changes` to better align with the existing CLI options.
